src/libopensc/card-cac1.c

```diff
@@ -0,0 +1,558 @@
+		goto done;
+
+	cert_type = val[0];
+	cert_ptr = val + 1;
+	cert_len = val_len - 1;
+
+	/* if the info byte is 1, then the cert is compressed, decompress it */
+	if ((cert_type & 0x3) == 1) {
+#ifdef ENABLE_ZLIB
+		r = sc_decompress_alloc(&priv->cache_buf, &priv->cache_buf_len,
+			cert_ptr, cert_len, COMPRESSION_AUTO);
+#else
+		sc_log(card->ctx, "CAC compression not supported, no zlib");
+		r = SC_ERROR_NOT_SUPPORTED;
+#endif
+		if (r)
+			goto done;
+		cert_ptr = val;
+	} else if (cert_len > 0) {
+		priv->cache_buf = malloc(cert_len);
+		if (priv->cache_buf == NULL) {
+			r = SC_ERROR_OUT_OF_MEMORY;
+			goto done;
+		}
+		priv->cache_buf_len = cert_len;
+		memcpy(priv->cache_buf, cert_ptr, cert_len);
+	}
+
+	/* OK we've read the data, now copy the required portion out to the callers buffer */
+	priv->cached = 1;
+	len = MIN(count, priv->cache_buf_len-idx);
+	memcpy(buf, &priv->cache_buf[idx], len);
+	r = len;
+done:
+	if (val)
+		free(val);
+	SC_FUNC_RETURN(card->ctx, SC_LOG_DEBUG_NORMAL, r);
+}
+
+/*
+ * CAC cards use SC_PATH_SELECT_OBJECT_ID rather than SC_PATH_SELECT_FILE_ID. In order to use more
+ * of the PKCS #15 structure, we call the selection SC_PATH_SELECT_FILE_ID, but we set p1 to 2 instead
+ * of 0. Also cac1 does not do any FCI, but it doesn't understand not selecting it. It returns invalid INS
+ * if it doesn't like anything about the select, so we always 'request' FCI for CAC1
+ *
+ * The rest is just copied from iso7816_select_file
+ */
+static int cac_select_file_by_type(sc_card_t *card, const sc_path_t *in_path, sc_file_t **file_out)
+{
+	struct sc_context *ctx;
+	struct sc_apdu apdu;
+	unsigned char buf[SC_MAX_APDU_BUFFER_SIZE];
+	unsigned char pathbuf[SC_MAX_PATH_SIZE], *path = pathbuf;
+	int r, pathlen, pathtype;
+	struct sc_file *file = NULL;
+	cac_private_data_t * priv = CAC_DATA(card);
+
+	assert(card != NULL && in_path != NULL);
+	ctx = card->ctx;
+
+	SC_FUNC_CALLED(ctx, SC_LOG_DEBUG_VERBOSE);
+
+	memcpy(path, in_path->value, in_path->len);
+	pathlen = in_path->len;
+	pathtype = in_path->type;
+
+	sc_debug(card->ctx, SC_LOG_DEBUG_VERBOSE,
+		 "path->aid=%x %x %x %x %x %x %x  len=%"SC_FORMAT_LEN_SIZE_T"u, path->value = %x %x %x %x len=%"SC_FORMAT_LEN_SIZE_T"u path->type=%d (%x)",
+		 in_path->aid.value[0], in_path->aid.value[1],
+		 in_path->aid.value[2], in_path->aid.value[3],
+		 in_path->aid.value[4], in_path->aid.value[5],
+		 in_path->aid.value[6], in_path->aid.len, in_path->value[0],
+		 in_path->value[1], in_path->value[2], in_path->value[3],
+		 in_path->len, in_path->type, in_path->type);
+	sc_debug(card->ctx, SC_LOG_DEBUG_VERBOSE, "file_out=%p index=%d count=%d\n",
+		 file_out, in_path->index, in_path->count);
+
+	/* Sigh, sc_key_select expects paths to keys to have specific formats. There is no override.
+	 * we have to add some bytes to the path to make it happy. A better fix would be to give sc_key_file
+	 * a flag that says 'no, really this path is fine'.  We only need to do this for private keys */
```

Comment:
Feel free to suggest a fix to `sc_key_select` in a separate PR.

Reply:
If I am right, the function in question here is `iso7816_select_file()`, which is something that should not take flags to behave differently than the ISO says for various partly incompatible cards.

The above function is mostly simplified and adjusted to the limitation that CAC1 has. But I updated the comment to reflect this.